### DISCUSSION TITLE
Advice for compiling for targets with old GLIBC

### DISCUSSION THREAD

**Discussion by myz-dev** [cross-rs/cross#1261] [Q&A]
Hi,
I am trying to cross compile my application (using `axum` + `tokio`) that has a lot of dependencies for the `i686-unknown-linux-gnu` target.
The table in the overview `README` shows, that the docker container used for the target's cross compilation supports GLIBC `2.31` onward
I am targeting a device that has `2.20` on it.
I think I should try to build my own docker container that has GLIBC `2.20` installed, but I fear I have a bigger problem that would not be solved by this approach:
Testing the compilation with the older image for `i686-unknown-linux-gnu` (ghcr.io/cross-rs/i686-unknown-linux-gnu:0.2.2), that has GLIB version `2.23` installed, I get following errors:

```
error: failed to run custom build command for `openssl-sys v0.9.80`

Caused by:
  process didn't exit successfully: `/target/debug/build/openssl-sys-115ccf982b69f88a/build-script-main` (exit status: 1)
  --- stderr
  /target/debug/build/openssl-sys-115ccf982b69f88a/build-script-main: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.29' not found (required by /target/debug/build/openssl-sys-115ccf982b69f88a/build-script-main)
  /target/debug/build/openssl-sys-115ccf982b69f88a/build-script-main: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.28' not found (required by /target/debug/build/openssl-sys-115ccf982b69f88a/build-script-main)
  /target/debug/build/openssl-sys-115ccf982b69f88a/build-script-main: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.25' not found (required by /target/debug/build/openssl-sys-115ccf982b69f88a/build-script-main)
warning: build failed, waiting for other jobs to finish...
```
And many errors like this one:
```
error: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.28' not found (required by /target/debug/deps/libtokio_macros-7a0d36fad7badacc.so)
   --> /home/myz/.cargo/registry/src/github.com-1ecc6299db9ec823/tokio-1.23.0/src/lib.rs:586:13
    |
586 |     pub use tokio_macros::select_priv_declare_output_enum;
    |         
```

 Does this mean that my dependencies inherently depend on `GLIBC_2.28` and I will not be able to compile my project using these crates for my target?
Is there anything I could do to run my software on this target that does not involve not using all dependencies that show errors during the cross compilation?

Reading [90846](https://github.com/rust-lang/rust/pull/90846) gave me some hope that I might be able to circumvent my issue with some kind of compilation parameter to use 'weak symbols', but to be honest I do not understand how to apply this information to my use case at all.
Is there anyone who would point me to the right direction on this issue?

**Comment by Emilgardis** [comment]
It seems you're hitting #724

Build-scripts are built and kept even if glibc changes, so a build-script compiled on the host can be carried over to the cross container with a lower glibc, causing #724

Now, for the glibc problem, you can solve this in various ways.

the simplest, is using cargo-zigbuild, integrated into our `zig` image and usable by

```toml
# Cross.toml
[target.i686-unknown-linux-gnu]
zig = "2.20"
```

another way, is piggy-backing on https://github.com/cross-rs/cross-toolchains and creating a image for `i686-unknown-linux-gnu` so that you can set glibc and even linux version. We don't currently have a config for it

**Reply by myz-dev** [reply]
Hi,
thank you very much for your quick reply!
The zig build route seems very promising, but unfortunately I could not make it work yet.
I have this `Cross.toml` file now:
```
[target.i686-unknown-linux-gnu]
pre-build = [
    "dpkg --add-architecture i386 && apt-get update && apt-get install --assume-yes libssl-dev:i386 && apt-get install --assume-yes zlib1g-dev && apt-get install --assume-yes zlib1g-dev:i386",
]
zig = "2.20"

```
Running `cross build --bin defect_map --target i686-unknown-linux-gnu` tells me that it could `not find directory of OpenSSL installation`. Checking the image that is written after the pre-build hook I can see with `dpkg -L libssl-dev` that the package is installed for the target architecture. 
Commenting out the `zig = "2.20"` line cross compiles, but of course with the same GLIBC version problem as before.
Do you know why the zig image is behaving differently in regards to the `libssl-dev` package? Why is it not found during the build process?

**Reply by Emilgardis** [reply]
ah, with openssl I think if you install `pkg-config` it will work

**Reply by Emilgardis** [reply]
otherwise, I'd recommend using the [`vendored` feature](https://docs.rs/openssl/latest/openssl/#vendored)

**Reply by myz-dev** [reply]
Hi,

again, thank you so much for your help. 
With the vendored option I could circumvent the `libssl-dev` problem, but I think I misunderstand something very fundamental.
I now get the same error for the `libgit2-dev` package. Of course in the pre-build hook I am installing the package with `apt-get install -y libgit2-dev:i386` (also tried it without the `:i386` architecture suffix). Also I have installed the `pkg-config` package (which was installed for the host architecture already, which I think is the only one needed. For good measure, I am also installing it with `DEBIAN_FRONTEND=noninteractive apt install --assume-yes pkg-config:i386`), but the openssl package and the git2 package are still not detected during the build step.
I have also tried passing environmental variables to the docker file that contain the directories the packages are installed to (checked by running the image interactively and running `dpkg -L libgit2-dev` for example). In the printed error messages of the failing build scripts I can see that the variables have been set and recognized, but the build scripts still can not locate the packages in question.
So I am very clueless to why this could be. Unfortunately I do not know much about this process, I am trying to learn still.

Forking the crate that uses gitlib2 (`built`) and enabling the `vendored-libgit2` feature in its `git2` dependency unfortunately did not solve the problem.
I will for now have to give up on the zig build and will try your second suggestion, creating my own image based on what I find here: https://github.com/cross-rs/cross-toolchains

Thanks

**Reply by Emilgardis** [reply]
hmm ok, unfortunate. [This](https://github.com/rust-cross/cargo-zigbuild/issues/127) and maybe setting `PKG_CONFIG_PATH=/usr/lib/i386-linux-gnu/pkgconfig/:${PKG_CONFIG_PATH}` would maybe be of use, but not sure


> Forking the crate that uses gitlib2 (built) and enabling the vendored-libgit2 feature in its git2 dependency unfortunately did not solve the problem.

just fyi, you could make your top crate depend (in `[build-dependencies]`) on `git2` and then specify the feature, instead of forking `built`, since features are enabled for every crate dependency across the build

**Reply by myz-dev** [reply]
Thank you for the hint, setting the path did not work unfortunately. I will learn about `PKG_CONFIG_PATH` now and see if I can fix the problem that way :-)

**Reply by myz-dev** [reply]
And: Thanks, just setting the feature in my main crate is great advice :-)
